Added update and delete for stories.

aks681 opened the pull request:
In this PR:

- Update and delete functionalities have been added for stories along with its corresponding tests.
- Validations have been added for Stories.

aks681 committed:
added story update

aks681 committed:
Added story delete

aks681 committed:
fixed errors

seanlip commented on core/domain/search_services.py:
```diff
@@ -27,6 +27,9 @@
 SEARCH_INDEX_COLLECTIONS = 'collections'
 
+# Name for the story search index.
+SEARCH_INDEX_STORIES = 'stories'
```
Hm, I don't think stories are meant to be searchable -- why do we need to change this file?

aks681 replied:
It was not needed, removed it.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -320,6 +324,188 @@ def update_story_contents_from_model(
+            exploration_id: str. The id of the exploration.
+
+        Returns:
+            Boolean. True if a node with the given exploration id is already
```
bool. Whether a node with the given exploration ID is already present.

aks681 replied:
Done.

seanlip commented on core/domain/story_services.py:
```diff
@@ -241,6 +246,226 @@ def save_new_story(committer_id, story):
+        change_list: list(dict). A change list to be applied to the given
+            story. Each entry in change_list is a dict that represents a
+            StoryChange.
+    object.
```
?

aks681 replied:
Done, edited the docstring.

seanlip commented on core/domain/story_services.py:
```diff
@@ -241,6 +246,226 @@ def save_new_story(committer_id, story):
+    object.
+
+    Returns:
+      Story. The resulting story domain object.
```
indent by 4 not 2

/cc @apb7 FYI -- linter seems to pass, but it shouldn't. Could you please address this, or file an issue to keep track?

aks681 replied:
Done.

apb7 replied:
@seanlip: The docstring params linter uses regexp and checks for missing params, missing param type, differing param type, missing returns, missing return type, differing return type, missing yield, missing raises, missing raises type and differing raises type. I do not think it checks for the indentation of docstrings. We might require another check for it. I will see this. Thanks!

apb7 replied:
I have added this to Issue #3905 under the Python checks.

seanlip commented on core/domain/story_services.py:
```diff
@@ -241,6 +246,226 @@ def save_new_story(committer_id, story):
+    # Validate that all explorations referenced by the story exist.
+    exp_ids = []
+    for node in story.story_contents.nodes:
+        exp_ids.append(node.exploration_id)
```
Might as well only append if not None, right?

aks681 replied:
Yeah, done.

seanlip commented on core/domain/story_services.py:
```diff
@@ -241,6 +246,226 @@ def save_new_story(committer_id, story):
+        committer_id: str. ID of the given committer.
+        story: Story. The story domain object to be saved.
+        commit_message: str. The commit message.
+        change_list: list(dict). List of changes applied to a story. Each
```
Do you think we could have this be list(StoryChange) instead? In general, better to work with domain objects rather than dicts. (I.e. convert to domain objects in the controller and pass the domain objects around.) Ditto for other functions in this file that use change dicts (and elsewhere in the codebase if you like, e.g. for explorations, but that's optional).

aks681 replied:
There are validations in _trusted_commit() in base_model/gae_models.py to make sure that changelist is a list of dicts. So, should I remove all those validations then?
Also, it is converted to its corresponding domain object at the beginning of apply_change_list().

seanlip replied:
Yes, if you are passing StoryChange domain objects, you don't need the dict validation. Instead you could have all the domain objects (StoryChange etc.) subclass BaseChange and check instead that the list contains a list of BaseChange objects.

And then you don't need to do any domain object conversions in apply_change_list, too.

(I feel like, since this pattern is becoming more prevalent with all the new models, we should probably do it properly rather than replicate the current suboptimal structure of passing dict blobs around.)

aks681 replied:
Done.

seanlip commented on core/domain/story_services.py:
```diff
@@ -241,6 +246,226 @@ def save_new_story(committer_id, story):
+    story_model.version = story.version
+    story_model.commit(committer_id, commit_message, change_list)
+    memcache_services.delete(_get_story_memcache_key(story.id))
+    index_story_given_id(story.id)
```
Probably not needed, as well as all the other search index stuff.

aks681 replied:
Ok, it was that initially, I only did the commit line, but then the tests were failing. So I wrote both these lines but it was because I did not have the memcache_services.delete() line before. Will remove this one.

aks681 replied:
Done.

seanlip commented on core/domain/story_services.py:
```diff
@@ -241,6 +246,226 @@ def save_new_story(committer_id, story):
+    - committer_id: str. The id of the user who is performing the update
+        action.
+    - story_id: str. The story id.
+    - change_list: list(dict). Each element in the list represents a valid
```
Modify to use list(StoryChange) instead -- see above.

aks681 replied:
Done.

seanlip commented on core/storage/story/gae_models.py:
```diff
@@ -77,22 +77,21 @@ def _trusted_commit(
         )
         story_commit_log_entry.story_id = self.id
         story_commit_log_entry.put_async()
+        super(StoryModel, self)._trusted_commit(
```
Can you explain why this was moved here? Just curious, and wondering whether we need to do it elsewhere too.

aks681 replied:
(answer for this and the comment below) I didn't test with the others, but when I manually tested with stories, the CommitLogEntry model that was created didn't increment version (didn't create an additional entry each time I did an edit) although it was incremented in _trusted_commit() in base_model. So, I thought it might be because of the put_async() that caused some kind of race condition, as entries in multiples of 2 were created (every other edit had a corresponding CommitLogEntry). So, I put the _trusted_commit() line after the CommitLogEntry line and explicitly incremented the version when creating the CommitLogEntry model without changing the variable itself (as this increment was done in _trusted_commit() ) and with this config, it started recording correctly.

seanlip replied:
Thanks for the explanation. That's not enough, though, I'm afraid -- we need to understand exactly (from an analytical point of view) why the change works and why the previous version of the code didn't work, and ensure that all models conform to a structure/pattern that works. Otherwise the current "recording correctly" might be a fluke. Could you investigate a bit more until you're absolutely sure about what the problem is?

Also, let me know if you need help, I'm happy to talk about this.

aks681 replied:
Done, changed it to put() as discussed. Also, confirmed that the CommitLogEntry records correctly after this change.

seanlip commented on core/storage/story/gae_models.py:
```diff
@@ -77,22 +77,21 @@ def _trusted_commit(
 
         story_commit_log_entry = StoryCommitLogEntryModel.create(
-            self.id, self.version, committer_id, committer_username,
+            self.id, self.version + 1, committer_id, committer_username,
```
Can you explain why this was changed? Just curious, and wondering whether we need to do it elsewhere too.

aks681 replied:
(explanation above) We might need to, I didn't test with explorations and collections yet. Will change it if it also has problems.

aks681 replied:
I checked and the existing code seems to work for collections and explorations.

aks681 replied:
Done, removed this increment.

seanlip reviewed (changes requested):
Thanks! Took a pass.

aks681 committed:
made initial changes

aks681 committed:
fixed comment

aks681 committed:
Added story validation

aks681 committed:
removed change dicts and modified _trusted_commit

aks681 commented:
@seanlip PTAL. Made the review changes.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -15,7 +15,9 @@
 """Domain objects relating to stories."""
 
 from constants import constants
+from core.domain import skill_services
```
story_domain should not import skill_services -- see comments on validation in the topic PR.

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -98,9 +100,36 @@ def __init__(self, change_dict):
+        """
+        story_change_dict = {}
+        story_change_dict['cmd'] = self.cmd
+        if hasattr(self, 'property_name'):
```
There's a fair amount of repetition here. Think it's worth having a class-level property

    OPTIONAL_ATTRIBUTE_NAMES = [...]

and then doing

    for attribute_name in self.OPTIONAL_ATTRIBUTE_NAMES:
        if hasattr(self, attribute_name):
            story_change_dict[attribute_name] = getattr(self, attribute_name)

perhaps?

This is optional, I'm happy with the current code too, it's a bit more explicit and sometimes more abstraction isn't the best thing anyway. So, just suggesting an alternative for your consideration, but ultimately it's your call. If you decide to go with this, make sure to do it for the other classes too.

aks681 replied:
Yeah, this makes it look cleaner. Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -182,6 +211,70 @@ def create_default_story_node(cls, node_id):
+                    'Expected exploration ID to be a string, received %s' %
+                    self.exploration_id)
+
+        if not isinstance(self.outline, basestring):
```
Just remembered something -- for any html strings (in any of the new domain objects) please make sure to pass them through html_cleaner.clean() on init(). This is important for security. See exp_domain.py for examples.

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -182,6 +211,70 @@ def create_default_story_node(cls, node_id):
+                raise utils.ValidationError(
+                    'Expected each prerequisite skill id to be a string, ' +
+                    'received %s' % skill_id)
+            if skill_services.get_skill_by_id(skill_id, False) is None:
```
Don't need to check this here. Ditto below.

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -182,6 +211,70 @@ def create_default_story_node(cls, node_id):
+                'Expected node ID to be a string, received %s' %
+                self.id)
+
+        if not isinstance(self.prerequisite_skill_ids, list):
```
Think about interactions between the attrs. For example I guess you don't want prerequisite_skill_ids and acquired_skill_ids to have a nonempty intersection, for any given node?

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -195,6 +288,30 @@ def __init__(self, story_nodes):
+                        (node for node in self.nodes
+                         if node.id == destination_node_id), None) is None:
+                    raise utils.ValidationError(
+                        'Expected all destination nodes to exist')
```
Should we also check that the story nodes form a connected acyclic graph?

aks681 replied:
I added a check to make sure a node doesn't point to itself.
Generally, nodes are unlocked when the prerequisites are acquired, right? So, in the check for cyclic, should that be taken into account also?

seanlip commented on core/domain/story_domain.py:
```diff
@@ -264,6 +381,49 @@ def __init__(
+                'Expected description to be a string, received %s'
+                % self.description)
+
+        if not isinstance(self.notes, basestring):
```
ditto -- this looks like html.

In general can we also annotate in the class attribute docstrings which fields are intended to be treated as html?

aks681 replied:
Done.

seanlip commented on core/domain/story_services.py:
```diff
@@ -225,6 +225,7 @@ def _create_story(committer_id, story, commit_message, commit_cmds):
         notes=story.notes,
         story_contents=story.story_contents.to_dict()
     )
+    commit_cmds = [commit_cmd.to_dict() for commit_cmd in commit_cmds]
```
Don't redefine vars. Use e.g. commit_cmd_dicts.

aks681 replied:
Done.

seanlip commented on core/domain/story_services.py:
```diff
@@ -387,22 +385,12 @@ def _save_story(committer_id, story, commit_message, change_list):
         ]
     }
     story_model.version = story.version
+    change_list = [change.to_dict() for change in change_list]
```
Don't redefine vars.

aks681 replied:
Done.

seanlip reviewed (changes requested):
Added some more notes. Thanks!

aks681 committed:
made review changes

aks681 committed:
Added tests and validations

seanlip commented on core/domain/story_domain.py:
```diff
@@ -272,8 +275,12 @@ def validate(self):
             if not isinstance(node_id, basestring):
                 raise utils.ValidationError(
-                    'Expected each node id to be a string, ' +
+                    'Expected each node id to be a string, '
```
node id --> destination node id

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -272,8 +275,12 @@ def validate(self):
                     'received %s' % node_id)
+            if node_id == self.id:
+                raise utils.ValidationError(
+                    'Expected destination node id list of a node to not have '
```
The wording here is a bit clumsy. How about: "The story node with ID {{node_id}} points to itself."?

Also, re:

> Generally, nodes are unlocked when the prerequisites are acquired, right? So, in the check for cyclic, should that be taken into account also?

Yes, I think so. One way you could do it is to simulate a user starting with "all prereq skills - all acquired skills" (presumably they've acquire the prereqs outside the story since there's nowhere else to do so). Then see if you can complete all the lessons using that as a basis, augmenting that list of skills along the way. I think we did something similar for collections, you might want to take a look at the code there too.

aks681 replied:
Done, I added the validation and a test that tests three cases for this, can you check if that's fine?

aks681 committed:
added additional validation

aks681 committed:
Fixed merge conflict

seanlip commented on core/domain/story_domain.py:
```diff
@@ -319,6 +318,83 @@ def validate(self):
+                    break
+            if node_unlocked:
+                unlocked_nodes.append(node)
+        print unlocked_nodes
```
drop this please

@apb7 FYI this sort of thing actually comes up rather regularly, might be worth including as a lint check (for stuff not in scripts/)?

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -319,6 +318,83 @@ def validate(self):
+        # based on the prerequisite skill ids calculated above.
+        for node in self.nodes:
+            node_unlocked = True
+            for skill_id in node.prerequisite_skill_ids:
```
Some suggestions for simplification:

- Create a helper method in this class called get_external_prerequisite_skill_ids() or similar. This computes the prerequisite skills that the nodes need that are not acquired from within the story, and returns them in a set.

- Replace line 321 with `simulated_skill_ids = get_external_prerequisite_skill_ids()` and add a comment above it saying "This set models the skill IDs acquired by a learner as they progress through the story.", since its definition appears to change over time.
- Replace this and the next four lines with: `if (set(node.prerequisite_skill_ids).issubset(simulated_skill_ids))`, and use similar constructs below as needed.

aks681 replied:
Changed the validations based on discussion.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -319,6 +318,83 @@ def validate(self):
+            # All of the initial unlocked nodes would be visited, but for
+            # consistency, that check is also done in this loop.
+            for node in unlocked_nodes:
+                node_visited = True
```
This should probably be called "node_is_accessible" since you haven't visited it yet.

aks681 replied:
Removed the flag variables.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -319,6 +318,83 @@ def validate(self):
+                    if skill_id not in prerequisite_skill_ids:
+                        node_visited = False
+                        break
+                if node_visited:
```
if node_is_accessible and node not in visited_nodes

aks681 replied:
Removed them.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -319,6 +318,83 @@ def validate(self):
             node.validate()
 
+        prerequisite_skill_ids = []
+        unlocked_nodes = []
```
Probably better to work with (primitive) node_ids than node objects

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -319,6 +318,83 @@ def validate(self):
+            if node_unlocked:
+                unlocked_nodes.append(node)
+        print unlocked_nodes
+        # The while loop is there so that, is a node is unlocked, but its
```
Can you define, in this comment or above the "unlocked_nodes" variable, what "unlocked" means? Does it mean: (i) all prerequisite skills are met; (ii) it is the destination node of some other node in the story; (iii) both of these?

aks681 replied:
Changed the structure of the validation.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -319,6 +318,83 @@ def validate(self):
+        if len(visited_nodes) < len(self.nodes):
+            raise utils.ValidationError('Expected all nodes to be reachable.')
+
+
```
Drop extra newline. /cc @apb7

aks681 replied:
Done.

seanlip commented on core/domain/story_domain_test.py:
```diff
@@ -171,6 +171,100 @@ def test_acquired_prerequisite_skill_intersection_validation(self):
             'Expected prerequisite skill ids and acquired skill ids '
             'to be mutually exclusive.')
 
+    def test_all_nodes_visited(self):
```
You'd probably want a few more tests for this: loops, disconnected graph, destination IDs are connected but skills don't work out, etc.

Try enumerating the various subcases, and then write a test for each one.

aks681 replied:
Done, added 4 tests that tests the various errors and a valid case.

aks681 committed:
Fixed merge conflict

aks681 committed:
Modified story validation

aks681 committed:
changed variables

seanlip commented on core/domain/story_domain.py:
```diff
@@ -32,12 +32,15 @@
 STORY_NODE_PROPERTY_OUTLINE = 'outline'
 STORY_NODE_PROPERTY_EXPLORATION_ID = 'exploration_id'
 
+STORY_CONTENTS_PROPERTY_STARTING_NODE_ID = 'starting_node_id'
```
Can we call it INITIAL_NODE_ID? Consistent with init_state_name in explorations, and it feels less like an action.

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -59,6 +62,8 @@ class StoryChange(object):
         STORY_NODE_PROPERTY_PREREQUISITE_SKILL_IDS, STORY_NODE_PROPERTY_OUTLINE,
         STORY_NODE_PROPERTY_EXPLORATION_ID)
 
+    STORY_CONTENTS_PROPERTIES = (STORY_CONTENTS_PROPERTY_STARTING_NODE_ID)
```
You will need a comma just before the last parenthesis if you want this to be treated as a tuple (and not a string).

https://wiki.python.org/moin/TupleSyntax

It is a surprise that this wasn't caught in tests. Before fixing this please write a test that contains a change that involves an element in this tuple and ensure that the test suite fails.

aks681 replied:
There is already a test in test_update_story_node_properties() in story_services_test that changes the starting node id and it passes.

seanlip replied:
That's weird. Why does the check here not trigger, then? https://github.com/oppia/oppia/pull/4988/files/4b22a75f039bc2a67ed214960c10ec30ab0e2a64..b79cee9522db6977e4a619f4e8c154fd125ae12b#diff-59604d490e874701fdc81e848603f084R111

EDIT: oh, I see. It's because you're comparing a string against the whole string. If you tried to use an invalid substring like 'TING_NODE' as the change cmd, I think it will go through.

aks681 replied:
Done, made it a tuple.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -305,6 +319,12 @@ def validate(self):
+                'Expected starting node id to be a string, received %s'
+                % self.starting_node_id)
+
+        starting_node_present = False
```
initial_node_is_present

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -317,82 +337,72 @@ def validate(self):
+                if not (set(destination_node.prerequisite_skill_ids).issubset(
+                        simulated_skill_ids)):
+                    raise utils.ValidationError(
+                        'Expected the user to have acquired all '
```
Probably worth saying which node and which skill.

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -317,82 +337,72 @@ def validate(self):
+        while len(nodes_queue) > 0:
+            current_node_id = nodes_queue.pop()
+            current_node_index = self.find_node_index(current_node_id)
+            # The following condition checks whether the destination node for
```
I don't see a "check" here? It just sets it?

aks681 replied:
Oh, that comment is misplaced. Had moved the check below, will move this.

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -317,82 +337,72 @@ def validate(self):
-
-        if len(visited_nodes) < len(self.nodes):
+        if False in is_node_visited:
             raise utils.ValidationError('Expected all nodes to be reachable.')
```
Again perhaps saying which nodes are not.

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -285,13 +297,15 @@ def validate(self):
+        self.starting_node_id = starting_node_id
         self.nodes = story_nodes
 
     def validate(self):
```
Btw are we validating uniqueness of all node IDs?

If not, might want to double check that we are validating absolutely everything that needs to be validated. Better to overvalidate if possible.

aks681 replied:
Done, added the validation, though node ids would be auto generated, so there is a good chance that they won't duplicate.

seanlip replied:
Hm, we should be certain. Thanks for adding the validation.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -317,82 +337,72 @@ def validate(self):
-
-    def get_node_by_id(self, node_id):
-        """Returns the node corresponding to the given id.
+    def find_node_index(self, node_id):
```
get_node_index perhaps, for consistency.

aks681 replied:
Done.

aks681 committed:
made review changes

seanlip commented on core/domain/story_domain.py:
```diff
@@ -32,7 +32,7 @@
 STORY_NODE_PROPERTY_EXPLORATION_ID = 'exploration_id'
 
-STORY_CONTENTS_PROPERTY_STARTING_NODE_ID = 'starting_node_id'
+INITIAL_NODE_ID = 'starting_node_id'
```
Shold change the hardcoded string too.

aks681 replied:
Done.

seanlip reviewed (approved):
All good, just one note.

Btw what is the format for node IDs and where are they generated? Even if they're generated in the frontend we should have a definition of their format (e.g. 6 random chars from a-z) and that definition should be validated in the backend.

aks681 committed:
Merge remote-tracking branch 'upstream/develop' into story_services

seanlip commented:
FYI there's a backend test failure.

aks681 committed:
Added validations and made review change

aks681 committed:
fixed errors

aks681 commented:
Done, added extra validation to check for valid node_id (using a next_node_id field in story_contents) and deleted a node id from all destination_node_id fields when its deleted.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -242,6 +242,10 @@ def validate(self):
                 'Expected node ID to be a string, received %s' %
                 self.id)
 
+        if 'node_' not in self.id:
```
Define a constant for this and refer to it where it's used (NODE_ID_PREFIX perhaps).

Also we should be validating more carefully. I.e. check that this string occurs at the _start_ and that it is followed by a number?

aks681 replied:
Done, the check for number is done in the validation of whether number is greater than next_node_id.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -297,16 +301,19 @@ def validate(self):
+            next_node_id: int. The index for the next node to be added to the
+                story.
         """
         self.starting_node_id = starting_node_id
```
I thought this was becoming the initial_node_id

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -428,7 +440,8 @@ def to_dict(self):
                 node.to_dict() for node in self.nodes
             ],
-            'starting_node_id': self.starting_node_id
+            'starting_node_id': self.starting_node_id,
```
shouldn't this be initial_node_id?

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -341,6 +348,11 @@ def validate(self):
                 initial_node_is_present = True
+            # Checks whether the number in the id of any node is greater than
+            # the value of next_node_id.
+            if int(node.id.replace('node_', '')) >= self.next_node_id:
```
For the first part of this clause, do something like node.get_number_from_node_id() instead. This behavior should be encapsulated in the Node object, you shouldn't be spreading details of how its id is generated outside where that information is needed.

aks681 replied:
Done, added the function to class StoryNode.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -297,16 +301,19 @@ def validate(self):
             story_nodes: list(StoryNode). The list of story nodes that are part
                 of this story.
             starting_node_id: str. The id of the starting node of the story.
+            next_node_id: int. The index for the next node to be added to the
```
So a node ID is a string. This should be number_for_next_node_id if it's an int, or it should be a string if it's actually meant to be a node ID. (Probably the latter and you can perhaps have a method for increment_next_node_id that does all the manipulation.)

aks681 replied:
Done, changed next_node_id to a string.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -341,6 +348,11 @@ def validate(self):
+                    'The node with id %s is out of bounds.' % node.id)
             node_id_list.append(node.id)
 
         if not initial_node_is_present:
```
Where's the validation for next_node_id?

aks681 replied:
Done, added validations.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -565,7 +581,7 @@ def create_default_story(cls, story_id):
                 StoryNode.create_default_story_node(
                     feconf.DEFAULT_INITIAL_NODE_ID)
```
Can we define this constant in this file, and have it use NODE_ID_PREFIX as a subcomponent?

In fact it may make sense to get rid of the constant altogether and have it as a local variable here (is it used elsewhere?) Reason is that, in feconf, you declare these constants so that it's easy for folks to change the config. But you really don't want people to arbitrarily change, given your system for naming nodes.

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -320,6 +612,196 @@ def update_story_contents_from_model(
+                'The node with id %s is not part of this story' % node_id)
+        if node_id == self.story_contents.starting_node_id:
+            raise ValueError(
+                'The node with id %s is the starting node for the story, change'
```
Missing space after "change"

aks681 replied:
Done.

aks681 committed:
made review changes

seanlip commented on core/domain/story_domain.py:
```diff
@@ -173,6 +176,26 @@ def __init__(
+
+        Args:
+            node_id: str. The id of the node.
+            external_id: bool. Whether the id to be decoded is the current
```
No, don't do this. Always pass in the node ID explicitly and always assume it is external. Also make this a `@classmethod` since it's not instance-specific.

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -242,7 +265,7 @@ def validate(self):
                 self.id)
 
-        if 'node_' not in self.id:
+        if NODE_ID_PREFIX != self.id[:len(NODE_ID_PREFIX)]:
```
This is insufficient. What if the node ID was node_kldjfskljlsfd?

You might consider having a NODE_ID_REGEX that is built out of NODE_ID_PREFIX among other things.

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -301,20 +324,33 @@ def validate(self):
+        """
+        # A node instance is required to access the function, and hence the
+        # first node is used.
+        current_number = self.nodes[0].get_number_from_node_id(
```
This isn't how you do it in python. Make get_number_from_node_id() a classmethod so that it's not bound to a specific instance, then call StoryNode.get_number_from_node_id(...)

Also this should probably be a classmethod in StoryNode as well since it pertains to node IDs which are basically a property of StoryNode. I.e. StoryNode.get_incremented_node_id(node_id). Then functions like story_contents.add_node() should call this, so that the incrementing happens automatically as part of the "add node" process.

aks681 replied:
Done, created the function.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -326,10 +362,23 @@ def validate(self):
-                % self.starting_node_id)
+                % self.initial_node_id)
+
+        if NODE_ID_PREFIX != self.initial_node_id[:len(NODE_ID_PREFIX)]:
```
This is insufficient -- see above. Also, given that this sort of logic is repeated in multiple places I recommend a classmethod along the lines of StoryNode.require_valid_node_id(string_to_test) -- remember the DRY concept. The validation that it is a string should be done outside that so that it's guaranteed that what's passed in is a string.

aks681 replied:
Done, added the function with regex checking and called that everywhere where node is was used.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -576,12 +627,14 @@ def create_default_story(cls, story_id):
             Story. The Story domain object with the default values.
         """
+        # Initial node id for a new story.
+        DEFAULT_INITIAL_NODE_ID = NODE_ID_PREFIX + '1'
```
initial_node_id = '%s1' % NODE_ID_PREFIX

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -576,12 +627,14 @@ def create_default_story(cls, story_id):
+                    DEFAULT_INITIAL_NODE_ID)
             ],
-            feconf.DEFAULT_INITIAL_NODE_ID, 2)
+            DEFAULT_INITIAL_NODE_ID, NODE_ID_PREFIX + '2')
```
StoryNode.get_incremented_node_id(initial_node_id)


(general principle: always try to pin things down to a single source of truth if they're related, rather than have two seemingly-independent pieces of data)

aks681 replied:
Done.

seanlip commented on core/domain/story_domain.py:
```diff
@@ -320,6 +665,197 @@ def update_story_contents_from_model(
+        self.story_contents.nodes[node_index].exploration_id = (
+            new_exploration_id)
+
+    def update_starting_node(self, new_initial_node_id):
```
update_initial_node

aks681 replied:
Done.

aks681 committed:
made review changes

seanlip commented:
LGTM -- see inline comment for future fix. Thanks! Merging.
